### ISSUE TITLE
[BUG] models:/name@alias does not use LoggedModel.artifact_location with --no-serve-artifacts

### ISSUE THREAD

**Issue by ebuchris** [mlflow/mlflow#25562]
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Other

### MLflow version

- Client: 3.15.2
- Tracking server: 3.15.2

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 22.04
- **Python version**:3.11

### Describe the problem

```
In MLflow 3, log_model(name=...) stores files at
<artifact-root>/<exp>/models/m-<model_id>/artifacts
not under the run tree
<artifact-root>/<exp>/<run_id>/artifacts/<name>.

The documented URI models:/<name>@<alias> still resolves through
ModelVersion.source (runs:/<run_id>/<name>), which becomes the run
filesystem path. With --no-serve-artifacts the client opens that path
and does not use LoggedModel.artifact_location.

Expected: download_artifacts("models:/name@alias") and
mlflow models serve -m "models:/name@alias" should load from
LoggedModel.artifact_location when that directory already exists
(no copy off NFS).

Actual:
- download_artifacts(alias) returns the empty run path (no MLmodel)
- LoggedModel.artifact_location has the real files
- ModelVersion.model_id is empty after get_model_version_by_alias
  (from_proto / to_proto TODO: model ID is not in the proto)

#23130 does not fix this. It only helps --serve-artifacts + runs:/ HTTP
fallback. Turning on --serve-artifacts would re-download into a temp dir
even though the files are already on the shared disk.

Related: #22141, #24535, #16979.
```

### Tracking information

```shell
mlflow server \
  --backend-store-uri sqlite:////tmp/mlflow-nfs-repro/db/mlflow.db \
  --default-artifact-root /tmp/mlflow-nfs-repro/artifacts \
  --no-serve-artifacts \
  --host 127.0.0.1 \
  --port 5000```


### Code to reproduce issue

<!-- PLEASE KEEP BACKTICKS AND CHECK PREVIEW -->
```
# 1. Start the server as in Tracking information.
# 2. Run: python reproduce.py  (same directory as this issue, or paste below)

from pathlib import Path
from urllib.parse import urlparse

import mlflow
from mlflow import MlflowClient
from mlflow.artifacts import download_artifacts
from sklearn.datasets import load_iris
from sklearn.linear_model import LogisticRegression

mlflow.set_tracking_uri("http://127.0.0.1:5000")
mlflow.set_experiment("alias-nfs-repro")

X, y = load_iris(return_X_y=True)
model = LogisticRegression(max_iter=200).fit(X, y)

with mlflow.start_run():
    info = mlflow.sklearn.log_model(model, name="model")

mlflow.register_model(info.model_uri, "alias_nfs_repro")
client = MlflowClient()
version = client.get_latest_versions("alias_nfs_repro")[0].version
client.set_registered_model_alias("alias_nfs_repro", "champion", version)

mv = client.get_model_version_by_alias("alias_nfs_repro", "champion")
logged = client.get_logged_model(info.model_id)
alias_path = Path(download_artifacts("models:/alias_nfs_repro@champion"))
logged_path = Path(urlparse(logged.artifact_location).path)

print("ModelVersion.model_id", mv.model_id)
print("ModelVersion.source", mv.source)
print("download_artifacts", alias_path, (alias_path / "MLmodel").is_file())
print("artifact_location", logged_path, (logged_path / "MLmodel").is_file())
```


### Stack trace

<!-- PLEASE KEEP BACKTICKS AND CHECK PREVIEW -->
```
No exception. download_artifacts("models:/name@alias") returns the run artifact path,
which does not contain MLmodel. The load/serve then fails later or copies from the
wrong tree.```

### Other info / logs

```
Layout on disk after log_model(name="model"):

  <artifact-root>/<exp_id>/models/m-<model_id>/artifacts/MLmodel   # real files
  <artifact-root>/<exp_id>/<run_id>/artifacts/model                # does not exist

Workaround that reads NFS in place (no re-download):

  mv = client.get_model_version_by_alias(name, alias)
  page = client.search_logged_models(
      experiment_ids=[client.get_run(mv.run_id).info.experiment_id],
      filter_string=f"source_run_id = '{mv.run_id}'",
      max_results=1,
  )
  model_id = page[0].model_id  # or page.models[0] depending on client
  local = client.get_logged_model(model_id).artifact_location```


### Willingness to contribute

No. I cannot contribute a bug fix at this time.

### What component(s) does this bug affect?

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

**Comment by jasperjonkhans** [comment]
Yo, I looked into this and the fix seems to have two parts: preserve model_id through SQL storage and protobuf conversion, then have ModelsArtifactRepository prefer the linked LoggedModel.artifact_location, with the current source URI as a fallback. Could a maintainer confirm whether that is the intended scope? If so, once the issue is marked ready, I’d be happy to work on it and add coverage for alias resolution with --no-serve-artifacts. Could you assign it to me at that point?

**Comment by kriscon-db** [comment]
Thanks for the report @ebuchris, and for the analysis @jasperjonkhans.

We dug into this on `master`. The `log_model(name=...)` → register → `models:/<name>@<alias>` flow resolves to the logged-model location, not the run tree:

```
# version source = models:/<model_id>
models:/rm@champ  ->  .../models/m-5ced.../artifacts   # LoggedModel.artifact_location
```

`register_model` sets the version `source` to `models:/<model_id>`, and `create_model_version` records `storage_location` as the linked `LoggedModel.artifact_location` — and that logic has been in place since MLflow 3.0 (#13211), so it applies to 3.15.2 as well. The only case that resolves to the run tree is a version whose `source` is an explicit `runs:/…` URI:

```
# version source = runs:/<run_id>/mymodel
models:/rm_runs@champ  ->  runs:/<run_id>/mymodel      # run tree
```

Since the standard flow produces a `models:/<model_id>` source, the version you are hitting this with likely has a `runs:/…` source.

@ebuchris could you share:
1. how the version was registered (the exact `log_model` / `register_model` / `create_model_version` call), and
2. the version's `source` (`client.get_model_version(name, version).source`)?

If a documented path is producing a `runs:/…` source for a logged model, that is the real bug and we will fix it; if it is an explicit `runs:/…` registration, `models:/<model_id>` is the URI that resolves to the logged-model artifacts.

**Comment by ebuchris** [comment]
HI,
Thank you for the response, I checked it and the bug was on my end after logging the model in the registering part.
Sorry,
Elad
